cgroup: support systemd properties via annotations

giuseppe opened the pull request:
add support for systemd properties passed through annotations in the
form:

"annotations": {
               "org.systemd.property.TimeoutStopUSec": "uint64 123456789",
                "org.systemd.property.CollectMode":"'inactive-or-failed'"
}

add a basic parser for the gvariant types to avoid a dependency on
glib.  The parser doesn't support complex types but it is just enough
for the types accepted by systemd.

Similar to https://github.com/opencontainers/runc/pull/2224

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

giuseppe commented:
@kolyshkin PTAL

rhatdan commented:
LGTM

rhatdan commented:
@kolyshkin Does this work for you?

giuseppe commented:
> @kolyshkin Does this work for you?

I am keeping in on hold as the runc patch wasn't merged yet.  Also there is an extra feature currently present in the runc PR that is not implemented for crun (convert *Sec properties to USec) yet.  I don't like the implicit conversion, so I am waiting how the discussion goes there before implementing it :-)

kolyshkin commented on src/libcrun/cgroup.c:
```diff
@@ -998,7 +1127,10 @@ int destroy_systemd_cgroup_scope (const char *scope, libcrun_error_t *err)
 static int
-libcrun_cgroup_enter_internal (runtime_spec_schema_config_linux_resources *resources, int cgroup_mode, char **path, const char *cgroup_path, int manager, pid_t pid, const char *id, libcrun_error_t *err)
+libcrun_cgroup_enter_internal (runtime_spec_schema_config_linux_resources *resources,
+                               json_map_string_string *annotations arg_unused, int cgroup_mode,
```
is the `arg_unused` here because of `HAVE_SYSTEMD` might be unset?

giuseppe replied:
yeah, it is unused only when not compiled with systemd

kolyshkin commented on src/libcrun/cgroup.c:
```diff
@@ -802,6 +904,33 @@ int enter_systemd_cgroup_scope (runtime_spec_schema_config_linux_resources *reso
+
+      for (i = 0; i < annotations->len; i++)
+        {
+          if (has_prefix (annotations->keys[i], "org.systemd.property."))
```
`s/"org.systemd.property."/SYSTEMD_PROPERTY_PREFIX/`

kolyshkin replied:
This is just a minor styling issue, but I'd rather have
```c
if (!has_prefix(annotations->keys[i], SYSTEMD_PROPERTY_PREFIX))
    continue
```

to improve readability and reduce the levels of indentation

kolyshkin commented on src/libcrun/cgroup.c:
```diff
@@ -690,8 +692,108 @@ systemd_job_removed (sd_bus_message *m, void *userdata, sd_bus_error *error arg_
+      if (UNLIKELY (sd_err < 0))
+        return crun_make_error (err, -sd_err, "sd-bus message append `%s`", name);
+    }
+
```
I'm sure you don't want to reimplement the entirety of gvariant parser here, yet a few notes:

1 gvariant assumes `int32` for value without an explicit type that can be parsed as a number. It might make sense to support it here.

2. I'd add checking of `endptr` from `strtoXX()` functions for being NULL (making sure there's no junk after the number).

3. There should be a whitespace after the type, i.e. `uint64123` is not valid but it is accepted by this parser.

At the very least I'd implement item 1.

kolyshkin commented:
Left a few comments, mostly nitpicks. Indeed we should wait until opencontainers/runc#2224 is merged.

giuseppe commented:
> Left a few comments, mostly nitpicks. Indeed we should wait until [opencontainers/runc#2224](https://github.com/opencontainers/runc/pull/2224) is merged.

thanks, I've addressed your comments.

rhatdan commented:
@giuseppe Is this ready to merge?

giuseppe commented:
> @giuseppe Is this ready to merge?

still waiting on the runc patch to make sure we support the same annotation.

Reviews for runc take a bit longer :-)

giuseppe commented:
the patch for runc got merged, ready for review

rhatdan commented:
LGTM
